scripts/safe_template_linter.py

```diff
@@ -1536,19 +1689,94 @@ def _check_filters(self, expression, context, has_page_default, results):
             elif (len(filters) == 2) and (filters[0] == 'n') and (filters[1] == 'js_escaped_string'):
                 # {x | n, js_escaped_string} is valid, if surrounded by quotes
-                pass
+                self._check_js_string_expression_in_quotes(mako_template, expression, results)
             else:
                 results.violations.append(ExpressionRuleViolation(
                     Rules.mako_invalid_js_filter, expression
                 ))
 
+    def _check_js_string_expression_in_quotes(self, mako_template, expression, results):
+        """
+        Checks that a Mako expression using js_escaped_string is surrounded by
+        quotes.
+
+        Arguments:
+            mako_template: The contents of the Mako template.
+            expression: A dict containing the start_index, end_index, and
+                expression (text) of the expression.
+            results: A list of results into which violations will be added.
+        """
+        parse_string = self._find_string_wrapping_expression(mako_template, expression)
+        if parse_string is None:
+            results.violations.append(ExpressionRuleViolation(
+                Rules.mako_js_missing_quotes, expression
+            ))
+
+    def _check_js_expression_not_with_html(self, mako_template, expression, results):
+        """
+        Checks that a Mako expression in a JavaScript context does not appear in
+        a string that also contains HTML.
+
+        Arguments:
+            mako_template: The contents of the Mako template.
+            expression: A dict containing the start_index, end_index, and
+                expression (text) of the expression.
+            results: A list of results into which violations will be added.
+        """
+        parse_string = self._find_string_wrapping_expression(mako_template, expression)
+        if parse_string is not None and re.search('[<>]', parse_string.string) is not None:
+            results.violations.append(ExpressionRuleViolation(
+                Rules.mako_js_html_string, expression
+            ))
+
+    def _find_string_wrapping_expression(self, mako_template, expression):
+        """
+        Finds the string wrapping the Mako expression if there is one.
+
+        Arguments:
+            mako_template: The contents of the Mako template.
+            expression: A dict containing the start_index, end_index, and
+                expression (text) of the expression.
+
+        Returns:
+            ParseString representing a scrubbed version of the wrapped string,
+            where the Mako expression was replaced with "${...}", if a wrapped
+            string was found.  Otherwise, returns None.
+        """
+        lines = StringLines(mako_template)
+        start_index = lines.index_to_line_start_index(expression['start_index'])
+        if 0 < expression['end_index']:
+            end_index = lines.index_to_line_end_index(expression['end_index'])
+        else:
+            return None
```

Comment:
Can the expression really not have an end_index? And if so, it's ok to not return a string?